otherlibs/systhreads/st_stubs.c

```diff
@@ -435,37 +430,27 @@ CAMLprim value caml_thread_cleanup(value unit)
 
 static void caml_thread_stop(void)
 {
-  caml_thread_t next;
-
   /* PR#5188, PR#7220: some of the global runtime state may have
      changed as the thread was running, so we save it in the
      curr_thread data to make sure that the cleanup logic
      below uses accurate information. */
   caml_thread_save_runtime_state();
 
-  next = Current_thread->next;
-
   /* The main domain thread does not go through [caml_thread_stop]. There is
      always one more thread in the chain at this point in time. */
-  CAMLassert(next != Current_thread);
+  CAMLassert(Active_thread->next != Active_thread);
 
-  caml_threadstatus_terminate(Terminated(Current_thread->descr));
-  caml_thread_remove_info(Current_thread);
-
-  /* FIXME: tricky bit with backup thread
-
-     Normally we expect the next thread to kick in and resume operation by
-     first setting Current_thread to the right TLS dec data. However it may
-     very well be that there's no runnable dec next (eg: next dec is
-     blocking.), so we set it to next for now to give a valid state to the
-     backup thread. */
-  Current_thread = next;
+  caml_threadstatus_terminate(Terminated(Active_thread->descr));
 
+  /* The following also sets Active_thread to a sane value in case the
+     backup thread does a GC before the domain lock is acquired
+     again. */
+  caml_thread_remove_info(Active_thread);
   caml_thread_restore_runtime_state();
 
   /* If no other OCaml thread remains, ask the tick thread to stop
      so that it does not prevent the whole process from exiting (#9971) */
-  if (All_threads == NULL) caml_thread_cleanup(Val_unit);
+  if (Active_thread == NULL) caml_thread_cleanup(Val_unit);
```

Comment:
Given the earlier comment about the main thread not going through `caml_thread_stop` and the assertion that there is at least one other thread, is it possible at all that `Active_thread` becomes NULL here?

Reply:
see https://github.com/ocaml/ocaml/pull/11271#discussion_r902642216